Interlinks Autocomplete: Mark II

jleandroperez opened the pull request:
### Details
In this PR we're implementing the core logic of Interlinks:

- **InterlinkProcessor:** In charge of determining when the Interlink Autocomplete must be shown / hidden
- **InterlinkViewController:** Renders Autocomplete suggestions above / below the cursor

**Note:** Nothing is expected to happen, at this stage, when you press over any row.

**Apologies about the size of the PR!!!** I've tried splitting this further, but testing without an actual UI makes no sense.

cc @eshurakov **Thank yoou sir!!!**

Ref. #914

---

### Test: Orientation
1. Open the Editor and add a note with 6 lines
2. Place the cursor in the 6th line
3. Type `[keyword` (with `keyword` being a word contained in any of your note's titles)

- [x] Verify Autocomplete shows up below the cursor's location
- [x] Verify that scrolling over the TextView below causes the Autocomplete window to follow along
- [x] Verify that if you scroll enough, the Autocomplete UI shows **below the Navigation Bar**
- [x] Verify that if your (updated) results yield just one row, the Autocomplete View doesn't jump above the keyword
- [x] Verify that if there are (way more) lines above, typing `[keyword` displays Autocomplete **above** the cursor

### Test: Filtering
1. Add a new note
2. In the new note's title, enter a word that's not present anywhere else. Suggestion `TEST999` perhaps?
3. Place the cursor in a newline
4. Type `[TEST999`

- [x] Verify Autocomplete does not show up (we intend to prevent circular references!)


### Test: Adaptivity
1. Open the Editor and type `[keyword` to trigger Autocomplete

- [x] Verify that rotating the device causes Autocomplete to disappear / reappear in the right location

### Test: UX
1. Open the Editor and type `[keyword` to trigger Autocomplete

- [x] Verify that a maximum of `3.5` rows always show up (we intend to avoid huge tables!!)
- [x] Verify Autocomplete reacts well to Dark / Light mode
- [ ] Verify Autocomplete reacts well to Dynamic Fonts
- [x] Verify that, on highlight, Autocomplete's First and Last rows get rounded corners

### Test: Dismissal
Please trigger the Autocomplete window, and verify it can be dismissed as follows:

- [x] Verify that pressing anywhere else in the TextView causes Autocomplete to get dismissed
- [x] Verify that pressing over any of the Navigation Bar buttons causes interlink to dismiss
- [x] Verify that pressing Back causes Interlink to get dismissed

### Release
These changes do not require release notes.

### Screenshots

<img width="200" src="https://user-images.githubusercontent.com/1195260/98828747-352e3300-2417-11eb-8aef-b8c276ed295e.png"> <img width="200" src="https://user-images.githubusercontent.com/1195260/98828764-38c1ba00-2417-11eb-8989-2e36be0312e6.png">

Commit:
New UIColor property

Commit:
UITextView: New Geometry API(s)

Commit:
UITextView: New Interlinking API(s)

Commit:
UITextView: New properties

Commit:
UIView: New Animations API(s)

Commit:
UIViewController: New containment API(s)

Commit:
Implements InterlinkProcessor

Commit:
Implements InterlinkViewController

Commit:
Update project.pbxproj

Commit:
Interlink: Reverts Link Insertion handlers

Commit:
InterlinkProcessor: Removes delegate

Commit:
InterlinkProcessor: Drops property

Commit:
InterlinkProcessor: Updates public API

Commit:
InterlinkProcessor: Renames API

Commit:
InterlinkProcessor: Adds new comments

Commit:
InterlinkViewController: Drops unused API

Commit:
InterlinkViewController: Missing cornerRadius

Commit:
Implements PassthruView

Commit:
InterlinkViewController: Cleanup Constraints mechanism

Commit:
InterlinkViewController: Shadow

Commit:
InterlinkViewController: Renames API

Commit:
InterlinkViewController: TableView Padding

Commit:
InterlinkProcessor: New delegate API

Commit:
Merge branch 'issue/914-interlinks-autocomplete-mk1' into issue/914-interlinks-autocomplete-mk2.1

Commit:
InterlinkProcessor: Fixing invocation

Commit:
Merge branch 'issue/914-interlinks-autocomplete-mk1' into issue/914-interlinks-autocomplete-mk2.1

Commit:
InterlinkProcessor: Fixes invocation

Commit:
InterlinkProcessor: Preventing overwork

Commit:
Merge branch 'issue/914-interlinks-autocomplete-mk1' into issue/914-interlinks-autocomplete-mk2.1

Commit:
InterlinkViewController: Preventing jumps

Commit:
InterlinkViewController: Updates calculation

Commit:
Editor: Dismissing Interlink on editionDidEnd

Commit:
Editor: Dismissing Interlink on present

Commit:
Merge branch 'issue/914-interlinks-autocomplete-mk1' into issue/914-interlinks-autocomplete-mk2.1

Commit:
InterlinkViewController: Adjusting colors

eshurakov commented on Simplenote/Classes/InterlinkProcessor.swift:
```diff
@@ -0,0 +1,179 @@
+
+
+
+// MARK: - InterlinkPresentationContextProvider
+//
+protocol InterlinkProcessorPresentationContextProvider: NSObjectProtocol {
+
+    /// View in the parent's hierarchy that should always appear above the Autocomplete UI
+    ///
+    func parentOverlayViewForInterlinkProcessor(_ processor: InterlinkProcessor) -> UIView
+
+    /// Parent TextView that is triggering the Autocomplete event
+    ///
+    func parentTextViewForInterlinkProcessor(_ processor: InterlinkProcessor) -> UITextView
+
+    /// Parent ViewController!
+    ///
+    func parentViewControllerForInterlinkProcessor(_ processor: InterlinkProcessor) -> UIViewController
+}
+
+
+// MARK: - InterlinkProcessorDatasource
+//
+protocol InterlinkProcessorDatasource: NSObjectProtocol {
+    var interlinkExcudedEntityID: NSManagedObjectID? { get }
+    var interlinkViewContext: NSManagedObjectContext { get }
+}
```
I was thinking, do we really need to have Context Provider and Data Source here? From what I can tell the only variable thing is `interlinkExcudedEntityID`, but it changes only when note changes. So maybe we can inject all of these dependencies instead of asking every time? I think it would simplify things and also less protocols to deal with. What do you think?

jleandroperez replied:
I've moved the `viewContext` into a required initializer parameter. However, the `excuded entity ID` is super cleaner to be fetched this way: the Note ID might change (IE if you press the + button).

I'll look into merging all the things into a single protocol (if I manage to come up with a proper name!!). Following up in another PR if you don't mind!  (Added a note in #914)

eshurakov commented on Simplenote/Classes/InterlinkProcessor.swift:
```diff
@@ -0,0 +1,179 @@
+private extension InterlinkProcessor {
+
+    func ensureInterlinkControllerIsOnScreen() {
+        if let _ = presentedViewController {
+            return
+        }
```
What do you think about checking `if presentedViewController != nil` or `guard presentedViewController == nil`. I feel it better expresses the intention.

jleandroperez replied:
I was gonna mention that `if let _ = ...` is **way faster** [if you need to check for null values](https://albinek.com/to-if-let-or-to-if-not-nil)... but [Google agrees with you!!](https://google.github.io/swift/).

Updating, thank you for the feedback sir!

eshurakov replied:
TIL that there is this different is performance between `if let _` and `!= nil` 👍

eshurakov commented on Simplenote/Classes/InterlinkProcessor.swift:
```diff
@@ -0,0 +1,179 @@
+    func presentInterlinkController() {
+        let interlinkViewController = InterlinkViewController()
+        interlinkViewController.attachWithAnimation(to: parentViewController, below: parentOverlayView)
+        self.presentedViewController = interlinkViewController
```
Nitpick: don't need `self` here :-)

jleandroperez replied:
Thank you!

eshurakov commented on Simplenote/Classes/UITextView+Simplenote.swift:
```diff
@@ -13,16 +14,41 @@ extension UITextView {
         }
 
-        return layoutManager.boundingRect(forGlyphRange: selectedRange, in: textContainer)
+        return text.indexFromLocation(selectedRange.location).flatMap { index in
+            text.interlinkKeyword(at: index)
+        }
```
Maybe we can do something like:
```
let range = Range(selectedRange, in: text)
let location = range.lowerBound
text.interlinkKeyword(at: location)
```

jleandroperez replied:
Sort of! `Range(selectedRange, in: text)` returns an optional. However, I love the idea of using an SDK API instead.

Thank you for the heads up!

eshurakov commented on Simplenote/Classes/UITextView+Simplenote.swift:
```diff
@@ -34,3 +60,56 @@ extension UITextView {
+    /// Returns the Bounding Rect for the specified `Range<String.Index>`
+    ///
+    func boundingRect(for range: Range<String.Index>) -> CGRect {
+        let nsRange = text.utf16NSRange(from: range)
```
I believe `NSRange(range, in: text)` should also work!

jleandroperez replied:
You're a genius, looks SO MUCH better!!

eshurakov commented on Simplenote/Classes/UITextView+Simplenote.swift:
```diff
@@ -34,3 +60,56 @@ extension UITextView {
+    ///
+    func locationInSuperviewForText(in range: Range<String.Index>) -> CGRect {
+        let rectInEditor = boundingRect(for: range)
+        return superview?.convert(rectInEditor, from: self) ?? rectInEditor
```
To convert this rect to window coordinates you can do: 
```
convert(rectInEditor, to: nil)
```
I think it's safer to convert it to window coordinates here and let the receiver to convert it from window coordinates to local coordinates.

eshurakov commented on Simplenote/Classes/UITextView+Simplenote.swift:
```diff
@@ -34,3 +60,56 @@ extension UITextView {
+            return .zero
+        }
+
+        return layoutManager.boundingRect(forGlyphRange: selectedRange, in: textContainer)
```
I think we need to convert `selectedRange` to glyph range first with 
```
open func glyphRange(forCharacterRange charRange: NSRange, actualCharacterRange actualCharRange: NSRangePointer?) -> NSRange
```

eshurakov commented on Simplenote/Classes/UIViewController+Simplenote.swift:
```diff
@@ -26,4 +26,21 @@ extension UIViewController {
+
+    /// Attaches a children ViewController (if needed) below the specified sibling view
+    ///
+    func attachWithAnimation(to parent: UIViewController, below siblingView: UIView) {
+        parent.view.insertSubview(view, belowSubview: siblingView)
+        parent.addChild(self)
+        view.fadeIn()
+    }
+
+    /// Detaches the receiver from its parent
+    ///
+    func detachWithAnimation() {
+        view.fadeOut { _ in
+            self.view.removeFromSuperview()
+            self.removeFromParent()
+        }
+    }
```
When adding a view controller as a child we need to call didMove after the view is added. Similar in removing.
I always forget the sequence and often consult with: 

https://developer.apple.com/library/archive/featuredarticles/ViewControllerPGforiPhoneOS/ImplementingaContainerViewController.html

:-)

eshurakov commented on Simplenote/Classes/SPNoteEditorViewController+Extensions.swift:
```diff
@@ -17,6 +17,13 @@ extension SPNoteEditorViewController {
+
+    /// Whenever a ViewController is presented, let's ensure Interlink is dismissed!
+    ///
+    override public func present(_ viewControllerToPresent: UIViewController, animated flag: Bool, completion: (() -> Void)? = nil) {
+        super.present(viewControllerToPresent, animated: flag, completion: completion)
+        interlinkProcessor.dismissInterlinkLookup()
+    }
```
Nice! Love the idea

jleandroperez replied:
Feels hacky but ... single spot to catch them all! Thank you!

eshurakov commented on Simplenote/Classes/InterlinkViewController.swift:
```diff
@@ -0,0 +1,184 @@
+private extension InterlinkViewController {
+
+    func setupRootView() {
+        view.frame = UIScreen.main.bounds
```
Do we need to set a frame here?

jleandroperez replied:
We do need to do this manually, **but perhaps** it's best for the snippet that instantiates this VC to initialize it's view size. I'll relocate this over to the InterlinkProcessor, thank you!

eshurakov commented on Simplenote/Classes/InterlinkViewController.swift:
```diff
@@ -0,0 +1,184 @@
+
+    /// Starts tracking ContentOffset changes in our sibling TextView
+    ///
+    func startObservingContentOffset(in textView: UITextView) {
+        kvoOffsetToken = textView.observe(\UITextView.contentOffset, options: [.old, .new]) { [weak self] (_, value) in
+            guard let oldY = value.oldValue?.y, let newY = value.newValue?.y, oldY != newY else {
+                return
+            }
+
+            self?.tableTopConstraint.constant += oldY - newY
+        }
+    }
```
I vote for listening for scrollViewDidScroll in the editor and passing it here. KVO gives me chills :-)

jleandroperez replied:
If it's okay with you, let's address that in a follow up PR! Added to my list of pendings in #914

eshurakov commented on Simplenote/Classes/InterlinkViewController.swift:
```diff
@@ -0,0 +1,184 @@
+    /// - Important: We'll always prefer either the "location above the cursor" whenever the **Maxed Out TableView** can be fit (that is: +3.5 results!)
+    ///
+    func calculateLocation(for height: CGFloat, around range: Range<String.Index>, in textView: UITextView) -> CGFloat {
+        let anchorFrame = textView.locationInSuperviewForText(in: range)
```
Do we need to convert it to a local view coordinates?

jleandroperez replied:
The TextView itself would be a sibling of this ViewController's view (both live in the same superview).

Since this VC is expected to assume the exact same size as it's parent, there should be no need for conversion (AFAIK!!), please correct me if I'm wrong!

eshurakov replied:
I would like to make as little assumptions as possible. I think safest approach would be if `textView.locationInSuperviewForText(in: range)` returns frame in Window coordinates and we would convert it back to our local coordinates here (similar to how we do it for keyboard frame).

eshurakov commented on Simplenote/Classes/InterlinkProcessor.swift:
```diff
@@ -0,0 +1,179 @@
+    @objc
+    func processInterlinkLookup() {
+        guard mustProcessInterlinkLookup,
+              let (_, keywordRange, keywordText) = parentTextView.interlinkKeywordAtSelectedLocation,
```
Idea: maybe we can already pass bounds for keywords instead of range?

jleandroperez replied:
Ranges are used for both, keyword positioning (And in a follow up PR) text replacement, I'm afraid.

Even if we do remove the positioning Range, we'd still need the full text range (for replacement purposes), soooo for consistency, I'd vote to keep them both the same. 

WDYT?

eshurakov replied:
I see! Let's see how it goes 👍

Commit:
InterlinkProcessor: Addressing Feedback

Commit:
UITextView+Simplenote: Addressing Feedback

Commit:
InterlinkViewController: Silences blur warning

Commit:
InterlinkViewController: Addressing feedback

Commit:
InterlinkProcessor: Addressing Feedback

Commit:
UITextView+Simplenote: Addressing Feedback

jleandroperez commented:
@eshurakov Thanks **a lot** for the awesome feedback!!. Addressed (most) of it, added some of them to my pendings list in #914 (will address them in a follow up PR).

Ready for another look. Thanks in advance!!

Commit:
Interlinks: Addressing feedback

eshurakov reviewed (approved):
Looks good! Thank you @jleandroperez

Commit:
Merge remote-tracking branch 'origin/develop' into issue/914-interlinks-autocomplete-mk2.1

jleandroperez commented:
Thanks a loot for reviewing this one!!
